pkg/object/mqttproxy/client.go

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mqttproxy
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/eclipse/paho.mqtt.golang/packets"
+	"github.com/megaease/easegress/pkg/logger"
+)
+
+const (
+	// Connected is MQTT client status of Connected
+	Connected = 1
+	// Disconnected is MQTT client status of Disconnected
+	Disconnected = 2
+
+	// Qos0 for "At most once"
+	Qos0 byte = 0
+	// Qos1 for "At least once
+	Qos1 byte = 1
+	// Qos2 for "Exactly once"
+	Qos2 byte = 2
+)
+
+type (
+	// ClientInfo is basic infomation for client
+	ClientInfo struct {
+		cid       string
+		username  string
+		password  string
+		keepalive uint16
+		will      *packets.PublishPacket
+	}
+
+	// Client represents a MQTT client connection in Broker
+	Client struct {
+		sync.Mutex
+
+		broker  *Broker
+		session *Session
+		conn    net.Conn
+
+		info   ClientInfo
+		status int
+		done   chan struct{}
+	}
+)
+
+func newClient(connect *packets.ConnectPacket, broker *Broker, conn net.Conn) *Client {
+	var will *packets.PublishPacket
+	if connect.WillFlag {
+		will = packets.NewControlPacket(packets.Publish).(*packets.PublishPacket)
+		will.Qos = connect.WillQos
+		will.TopicName = connect.WillTopic
+		will.Retain = connect.WillRetain
+		will.Payload = connect.WillMessage
+		will.Dup = connect.Dup
+	}
+
+	info := ClientInfo{
+		cid:       connect.ClientIdentifier,
+		username:  connect.Username,
+		password:  string(connect.Password),
+		keepalive: connect.Keepalive,
+		will:      will,
+	}
+	client := &Client{
+		broker: broker,
+		conn:   conn,
+		info:   info,
+		status: Connected,
+		done:   make(chan struct{}),
+	}
+	return client
+}
+
+func (c *Client) readLoop() {
+	defer c.close()
+	keepAlive := time.Duration(c.info.keepalive) * time.Second
+	timeOut := keepAlive + keepAlive/2
+	for {
+		select {
+		case <-c.done:
+			return
+		default:
+		}
+
+		if keepAlive > 0 {
+			if err := c.conn.SetDeadline(time.Now().Add(timeOut)); err != nil {
+				logger.Errorf("mqtt.readLoop: set read timeout failed, err:%s", c.info.cid)
+			}
+		}
+
+		packet, err := packets.ReadPacket(c.conn)
+		if err != nil {
+			logger.Errorf("mqtt.readLoop: client %s read packet failed, err:%v", c.info.cid, err)
+			return
+		}
+		err = c.processPacket(packet)
```

Comment:
It's better to distinguish different errors, for example

1. Server receives a DISCONN packet, it should disconnect the network connection as per [1], it's a normal operation, shouldn't be wrap as an error.
2. Server should report a protocol violation error (such as double connect), and handle it according to [2]. In our implementation, we'd better report error to logs in error level and disconnect network connection (your current implementation only report it in a debug level log)

[1] http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718090
[2] http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718110

Reply:
Thanks a lot for your advice! I will fix that!